<root main_tree_to_execute="BehaviorTree">
    <!-- ////////// -->
    <BehaviorTree ID="BehaviorTree">
        <Sequence>
            <Condition ID="RobotPret"/>
            <RealParallel>
                <SubTree ID="Tirette" y_mul="y_mul" theta_mul="theta_mul" camp="camp"/>
                <SubTree ID="Minuterie" palet_vert="palet_vert" palets_rouge="palets_rouge" palet_pente="palet_pente" />
                <SubTree ID="Yeux" camp="camp"/>
                <SubTree ID="Trier les palets" palet_vert="palet_vert" palets_rouge="palets_rouge" palet_pente="palet_pente" y_mul="y_mul" theta_mul="theta_mul" />
            </RealParallel>
        </Sequence>
    </BehaviorTree>
    <!-- ////////// -->
    <BehaviorTree ID="Tirette">
        <ForceSuccess>
            <Sequence>
                <Condition ID="TiretteLaserPresente"/>
                <Action ID="Arreter le robot" />
                <Action ID="DefinirLeScore" score="0"/>
                <Action ID="ArreterLaMinuterie"/>
                <Fallback>
                    <Sequence>
                        <Condition ID="LaVoieEstLibre-Gauche"/>
                        <Action ID="DefinirPosition" theta="0" x="-0.18" y="-1.18"/>
                        <SetBlackboard value="violet" output_key="camp"/>
                        <SetBlackboard value="-1.0" output_key="y_mul"/>
                        <SetBlackboard value="-1.0" output_key="sym_theta"/>
                        <Action ms="1000" ID="Attendre"/>
                    </Sequence>
                    <Sequence>
                        <Action ID="DefinirPosition" theta="0" x="-0.18" y="1.18"/>
                        <SetBlackboard value="jaune" output_key="camp"/>
                        <SetBlackboard value="1.0" output_key="y_mul"/>
                        <SetBlackboard value="1.0" output_key="theta_mul"/>
                        <Action ms="1000" ID="Attendre"/>
                    </Sequence>
                </Fallback>
            </Sequence>
        </ForceSuccess>
    </BehaviorTree>
    <!-- ////////// -->
    <BehaviorTree ID="Minuterie">
        <ForceSuccess>
            <Fallback>
                <Sequence>
                    <Inverter>
                        <Condition ID="TiretteLaserPresente"/>
                    </Inverter>
                    <Inverter>
                        <Condition ID="MinuterieDemarree"/>
                    </Inverter>
                    <Action seconds="88" ID="LancerLaMinuterie"/>
                    <SetBlackboard value="1" output_key="palet_vert"/>
                    <SetBlackboard value="1" output_key="palets_rouge"/>
                    <SetBlackboard value="1" output_key="palet_pente"/>
                </Sequence>
                 <Sequence>
                    <Inverter>
                        <Condition ID="MinuterieNonEcoulee"/>
                    </Inverter>
                    <SetBlackboard value="0" output_key="palet_vert"/>
                    <SetBlackboard value="0" output_key="palets_rouge"/>
                    <SetBlackboard value="0" output_key="palet_pente"/>
                    <Action ID="Arreter le robot"/>
                </Sequence>
            </Fallback>
        </ForceSuccess>
    </BehaviorTree>
    <!-- ////////// -->
    <BehaviorTree ID="Trier les palets">
        <ForceSuccess>
            <Fallback>
                <BlackboardCheckString return_on_mismatch="FAILURE" value_A="{palet_vert}" value_B="1">
                    <Sequence>
                        <Action ID="Aller" theta="90"  x="0.30" y="1.15" y_mul="{y_mul}" theta_mul="{theta_mul}" comment="contourner les palets par l'extérieur"/>
                        <Action ID="Aller" theta="180" x="0.30" y="1.05" y_mul="{y_mul}" theta_mul="{theta_mul}" comment="se placer face aux palets"/>
                        <Action ID="Aller" theta="180" x="0.05" y="1.05" y_mul="{y_mul}" theta_mul="{theta_mul}" comment="capturer le palet vert"/>
                        <Action ID="Aller" theta="270" x="0.05" y="1.05" y_mul="{y_mul}" theta_mul="{theta_mul}" comment="tourner en directin de la zone de tri"/>
                        <Action ID="Aller" theta="270" x="0.05" y="1.25" y_mul="{y_mul}" theta_mul="{theta_mul}" comment="avancer dans la zone de tri"/>
                        <Action ID="Aller" theta="180" x="0.05" y="1.25" y_mul="{y_mul}" theta_mul="{theta_mul}" comment="se tourner en direction de la case verte"/>
                        <Action ID="Aller" theta="180" x="-0.25" y="1.25" y_mul="{y_mul}" theta_mul="{theta_mul}" comment="déposer le palet dans la case verte"/>
                        <Action ID="Aller" theta="180" x="0.1" y="1.25" y_mul="{y_mul}" theta_mul="{theta_mul}" comment="reculer pour libérer le palet"/>
                        <Action ID="AjouterAuScore" score="6"/>
                        <SetBlackboard value="0" output_key="palet_vert"/>
                    </Sequence>
                </BlackboardCheckString>
                <BlackboardCheckString return_on_mismatch="FAILURE" value_A="{palets_rouge}" value_B="1">
                    <Sequence>
                        <Action ID="Aller" theta="90" x="0.1" y="1.05" y_mul="{y_mul}" theta_mul="{theta_mul}" comment="remonter sur la ligne des palets"/>
                        <Action ID="Aller" theta="180" x="0.1" y="1.05" y_mul="{y_mul}" theta_mul="{theta_mul}" comment="se tourner vers les palets rouges"/>
                        <Action ID="Aller" theta="180" x="-0.55" y="1.05" y_mul="{y_mul}" theta_mul="{theta_mul}" comment="capturer les deux palets"/>
                        <Action ID="Aller" theta="270" x="-0.55" y="1.05" y_mul="{y_mul}" theta_mul="{theta_mul}" comment="se tourner en direction de la case rouge"/>
                        <Action ID="Aller" theta="270" x="-0.55" y="1.25" y_mul="{y_mul}" theta_mul="{theta_mul}" comment="pousser les palets au fond de la case rouge"/>
                        <Action ID="Aller" theta="270" x="-0.55" y="1.05" y_mul="{y_mul}" theta_mul="{theta_mul}" comment="libérer les palets"/>
                        <Action ID="AjouterAuScore" score="12"/>
                        <SetBlackboard value="0" output_key="palets_rouge"/>
                    </Sequence>
                </BlackboardCheckString>
                <BlackboardCheckString return_on_mismatch="FAILURE" value_A="{palet_pente}" value_B="1">
                    <Sequence>
                        <Action ID="Aller" theta="90" x="0.8" y="1.2" y_mul="{y_mul}" theta_mul="{theta_mul}" comment="aller au pied de la balance"/>
                        <Action ID="AjouterAuScore" score="12"/>
                        <SetBlackboard value="0" output_key="palet_pente"/>
                    </Sequence>
                </BlackboardCheckString>
            </Fallback>
        </ForceSuccess>
    </BehaviorTree>
    <!-- ////////// -->
    <BehaviorTree ID="Yeux">
        <RealParallel>
            <Sequence>
                <Action message="content.gif" repetition="1" defaut="1" ID="AfficheYeux" fps="2"/>
                <Action ms="10000" ID="Attendre"/>
                <Action message="carre.gif" repetition="1" defaut="1" ID="AfficheYeux" fps="10"/>
                <Action ms="10000" ID="Attendre"/>
                <Action message="folie.gif" repetition="1" defaut="1" ID="AfficheYeux" fps="2"/>
                <Action ms="10000" ID="Attendre"/>
                <Action message="perdu.gif" repetition="1" defaut="1" ID="AfficheYeux" fps="4"/>
                <Action ms="10000" ID="Attendre"/>
                <Action message="stardust.gif" repetition="1" defaut="1" ID="AfficheYeux" fps="1"/>
                <Action ms="10000" ID="Attendre"/>
            </Sequence>
            <ForceSuccess>
                <ReactiveFallback>
                    <Sequence>
                        <Inverter>
                            <Condition ID="RobotEnBonneSante"/>
                        </Inverter>    
                        <Action message="CPU !!!" repetition="1" defaut="0" ID="AfficheYeux" fps="20"/>
                    </Sequence>
                    <Sequence>
                        <Condition ID="TiretteLaserPresente"/>
                        <Action message="{camp}" repetition="1" defaut="0" ID="AfficheYeux" fps="40"/>
                    </Sequence>
                    <Sequence>
                        <Condition ID="MinuterieDemarree"/>
                        <Condition ID="MinuterieNonEcoulee"/>
                        <Action ms="10000" ID="Attendre"/>
                        <Action ID="RecupererLeScore" score="{score}"/>
                        <Action message="{score}" repetition="1" defaut="0" ID="AfficheYeux" fps="30"/>
                    </Sequence>
                    <Sequence>
                        <Inverter>
                            <Condition ID="MinuterieNonEcoulee"/>
                        </Inverter>    
                        <Action ms="5000" ID="Attendre"/>
                        <Action ID="RecupererLeScore" score="{score}"/>
                        <Action message="{score}" repetition="1" defaut="0" ID="AfficheYeux" fps="35"/>
                    </Sequence>
                </ReactiveFallback>
            </ForceSuccess>
        </RealParallel>
    </BehaviorTree>
    <!-- ////////// -->
    <TreeNodesModel>
        <Control ID="RealParallel" />
        <Action ID="AfficheYeux">
            <input_port default="0" name="defaut">1 = message par défaut</input_port>
            <input_port default="4" name="fps">Frame per second (image animée)</input_port>
            <input_port name="message">Message ou nom de l'image</input_port>
            <input_port default="1" name="repetition">Nombre d'itération</input_port>
        </Action>
        <Action ID="AjouterAuScore">
            <input_port name="score"/>
        </Action>
        <Action ID="Aller">
            <input_port name="theta"/>
            <input_port name="x"/>
            <input_port name="y"/>
            <input_port name="y"/>
            <input_port name="x_mul"/>
            <input_port name="y_mul"/>
            <input_port name="theta_mul"/>
            <input_port name="comment"/>
        </Action>
        <Action ID="ArreterLaMinuterie"/>
        <Action ID="Attendre">
            <input_port name="ms"/>
        </Action>
        <Action ID="DefinirLeScore">
            <input_port name="score"/>
        </Action>
        <Action ID="DefinirPosition">
            <input_port name="theta"/>
            <input_port name="x"/>
            <input_port name="y"/>
        </Action>
        <Action ID="RecupererLeScore">
            <output_port name="score"/>
        </Action>
        <SubTree ID="Initialisation"/>
        <Action ID="Arreter le robot" />
        <Condition ID="TiretteLaserPresente"/>
        <Condition ID="LaVoieEstLibre-Devant"/>
    <Condition ID="LaVoieEstLibre-Gauche"/>
   <Condition ID="LaVoieEstLibre-Droite"/>
    <Condition ID="LaVoieEstLibre-Derriere"/>

        <Action ID="LancerLaMinuterie">
            <input_port name="seconds"/>
        </Action>
        <SubTree ID="Minuterie"/>
        <Condition ID="MinuterieDemarree"/>
        <Condition ID="MinuterieNonEcoulee"/>
        <Condition ID="RobotPret"/>
        <Condition ID="ScoreInferieurA">
            <input_port name="score"/>
        </Condition>
        <SubTree ID="Trier les palets"/>
        <SubTree ID="Yeux"/>
    </TreeNodesModel>
    <!-- ////////// -->
</root>

